### ISSUE TITLE
simplified, self-contained Logger

### ISSUE THREAD

**Issue by pohly** [kubernetes/klog#360]
/kind feature

**Describe the solution you'd like**

In Kubernetes, we currently use the traditional klog output handling code, with all the cruft that has accumulated there. Since the deprecation of klog flags, none of that code does anything anymore, but we also cannot remove it because it is part of klog/v2.

What would be useful for Kubernetes would be a klogr Logger which:
- has its own, simple output handling, with support for buffering of info messages just as in zapr (new feature)
- support vmodule (feature parity)
- same output format (no usability changes)

The restructuring of klog into separate packages for vmodule and formatting should make the implementation of such a logger relatively easy.

**Comment by pohly** [comment]
/assign

A first implementation is showing signs of live...

**Comment by pohly** [comment]
Actually, there is already k8s.io/klog/v2/textlogger. I just reimplemented it in a slightly different (and better) way... :sweat_smile: 

I'll update it instead of adding a new logger.

**Comment by pohly** [comment]
The solution that I came up with installs textlogger with `klog.SetLoggerWithOptions(logger, klog.ContextualLogger(contextualLoggingEnabled))`. That means that `klog.Background().Info("msg")` will be handled entirely by the textlogger. Together with some optimizations, that increases encoding performance (= format, then write to io.Discard) for structured logging:

```
Encoding/kind-worker-kubelet/structured-36          318k ± 1%      364k ± 0%   +14.52%
```

But this approach has one drawback: a traditional `klog.Infof` first gets handled by klog, then the textlogger's `Info` is called to output the message string. The result is a drop in performance:

```
Encoding/kind-worker-kubelet/printf-36              642k ± 0%      228k ± 0%   -64.42%
```

In other words, such a switch only makes sense once most of the log calls in a component have been converted to structured logging.

**Comment by pohly** [comment]
Writing with an info buffer definitely improves performance:

```
Writing/tmp-files/structured/single-stream-36          106k ± 0%      132k ± 1%   +25.01%
Writing/tmp-files/structured/split-stream-36           105k ± 1%      245k ± 1%  +132.94%
```

**Comment by pohly** [comment]
PR #362 contains the changes for split output. The buffering of the info stream is implemented in the io.Writer used by k8s.io/component-base/logs. I don't have a PR open for that yet (depends on https://github.com/kubernetes/kubernetes/pull/114609).

**Comment by pohly** [comment]
> PR https://github.com/kubernetes/klog/pull/362 contains the changes for split output.

Actually, I had used a different approach for that there: by looking at the first byte of a write, the `io.Writer` implementation can determine whether it is an info or error message and redirect the write accordingly. No changes needed in klog for that...
